build-tools/packages/build-cli/test/commands/typetests.test.ts

```diff
@@ -2,11 +2,162 @@
+		devDependencies: {
+			"test-package-previous": "npm:test-package@2.0.0-internal.4.0.0",
+			"another-dependency": "^1.0.0",
+		},
+		typeValidation: {
+			broken: {
+				"broken-package": {
+					backCompat: false,
+					forwardCompat: false,
+				},
+			},
+			disabled: true,
+		},
+	};
+}
 
 describe("typetests tests", () => {
+	describe("applyTypeTestVersionOptions", () => {
+		describe("VersionOptions.Clear", () => {
+			it("removes previous test package dependency", () => {
+				const pkg = packageWithTypeValidation();
+				applyTypeTestVersionOptions(pkg, VersionOptions.Clear);
+				expect(pkg.devDependencies?.["test-package-previous"]).not.to.exist;
+			});
+
+			it("removes previous test package dependency when type tests are disabled", () => {
+				const pkg = packageWithTypeValidationDisabled();
+				applyTypeTestVersionOptions(pkg, VersionOptions.Clear);
+				expect(pkg.devDependencies?.["test-package-previous"]).not.to.exist;
+			});
+		});
+
+		describe("VersionOptions.ClearIfDisabled", () => {
+			it("leaves dependency when type tests are not disabled", () => {
+				const pkg = packageWithTypeValidation();
+				applyTypeTestVersionOptions(pkg, VersionOptions.ClearIfDisabled);
+				expect(pkg.devDependencies?.["test-package-previous"]).to.equal(
+					"npm:test-package@2.0.0-internal.4.0.0",
+				);
+			});
+
+			it("removes previous test package dependency when type tests are disabled", () => {
+				const pkg = packageWithTypeValidationDisabled();
+				applyTypeTestVersionOptions(pkg, VersionOptions.ClearIfDisabled);
+				expect(pkg.devDependencies?.["test-package-previous"]).not.to.exist;
+			});
+		});
+
+		describe("VersionOptions.Previous", () => {
+			it("sets previous version", () => {
+				const pkg = packageWithTypeValidation();
+				applyTypeTestVersionOptions(pkg, VersionOptions.Previous);
+				expect(pkg.devDependencies?.["test-package-previous"]).to.equal(
+					"npm:test-package@2.0.0-internal.5.0.0",
+				);
+			});
+
+			it("sets previous version even without typeValidation node", () => {
+				const pkg = packageMinimal();
+				applyTypeTestVersionOptions(pkg, VersionOptions.Previous);
+				expect(pkg.devDependencies?.["test-package-previous"]).to.equal(
+					"npm:test-package@2.0.0-internal.5.0.0",
+				);
+			});
+
+			it("removes previous test package dependency when type tests are disabled", () => {
+				const pkg = packageWithTypeValidationDisabled();
+				applyTypeTestVersionOptions(pkg, VersionOptions.Previous);
+				expect(pkg.devDependencies?.["test-package-previous"]).not.to.exist;
+			});
+		});
+	});
+
+	describe("resetBrokenTests", () => {
+		it(`ignores packages with no typeValidation node`, () => {
+			const pkg = packageMinimal();
+			resetBrokenTests(pkg, true);
+			expect(pkg.typeValidation).to.not.exist;
+		});
+
+		it(`resets package with type tests`, () => {
+			const pkg = packageWithTypeValidation();
+			resetBrokenTests(pkg, true);
+			expect(pkg.typeValidation?.broken).to.be.empty;
```

Comment:
personally, I prefer `assert.deepEqual(pkg.typeValidation?.broken, [])`

Makes the empty case not special compared to other cases, and also is more obvious about what it does. For example as someone who doesn't use chai a lot, I'm not confident that the assert as you have it would error on undefined or empty objects, empty maps etc, while asserting equal to [] is very clear.

Reply:
`broken` is an object. Should it be `assert.deepEqual(pkg.typeValidation?.broken, {})`?

Reply:
yes, and thats a nice reason to use deepEqual instead of to.be.empty: it makes it more clear what its doing :)